flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/failure/Failure.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.adaptive.failure;
+
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.scheduler.exceptionhistory.ExceptionHistoryEntry;
+import org.apache.flink.runtime.scheduler.exceptionhistory.RootExceptionHistoryEntry;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+/** Failure object. */
+public abstract class Failure {
+    private final Throwable cause;
+    private final long timestamp;
+
+    public Failure(Throwable cause) {
+        this.cause = cause;
+        this.timestamp = System.currentTimeMillis();
```

Comment:
This timestamp is reported to the user as time when the failure appeared. Is this okay, or do we want to have something different?

Reply:
Not really, the DefaultlScheduler does the same thing. Unless the timestamp would be reported by the TM, there is not much we can do